core/slim_sim.cpp

```diff
@@ -6191,13 +6202,111 @@ void SLiMSim::WritePopulationTable(tsk_table_collection_t *p_tables)
 	}
 }
 
+void SLiMSim::WriteTreeSequenceMetadata(tsk_table_collection_t *p_tables)
+{
+    int ret = 0;
+
+    //////
+    // Top-level (tree sequence) metadata:
+    // In the future, we might need to *add* to the metadata *and also* the schema,
+    // leaving other keys that might already be there.
+    // But that's being a headache, so we're skipping it.
+    nlohmann::json metadata;
+	metadata["SLiM"]["model_type"] = (ModelType() == SLiMModelType::kModelTypeWF) ? "WF" : "nonWF";
+	metadata["SLiM"]["generation"] = Generation();
+	metadata["SLiM"]["file_version"] = SLIM_TREES_FILE_VERSION;
+    if (spatial_dimensionality_ == 0) {
+        metadata["SLiM"]["spatial_dimensionality"] = "";
+    } else if (spatial_dimensionality_ == 1) {
+        metadata["SLiM"]["spatial_dimensionality"] = "x";
+    } else if (spatial_dimensionality_ == 2) {
+        metadata["SLiM"]["spatial_dimensionality"] = "xy";
+    } else {
+        metadata["SLiM"]["spatial_dimensionality"] = "xyz";
+    }
+    if (periodic_x_ & periodic_y_ & periodic_z_) {
+        metadata["SLiM"]["spatial_periodicity"] = "xyz";
+    } else if (periodic_x_ & periodic_y_) {
+        metadata["SLiM"]["spatial_periodicity"] = "xy";
+    } else if (periodic_x_ & periodic_z_) {
+        metadata["SLiM"]["spatial_periodicity"] = "xz";
+    } else if (periodic_y_ & periodic_z_) {
+        metadata["SLiM"]["spatial_periodicity"] = "yz";
+    } else if (periodic_x_) {
+        metadata["SLiM"]["spatial_periodicity"] = "x";
+    } else if (periodic_y_) {
+        metadata["SLiM"]["spatial_periodicity"] = "y";
+    } else if (periodic_z_) {
+        metadata["SLiM"]["spatial_periodicity"] = "z";
+    } else {
+        metadata["SLiM"]["spatial_periodicity"] = "";
+    }
+	metadata["SLiM"]["separate_sexes"] = sex_enabled_ ? true : false;
+	metadata["SLiM"]["nucleotide_based"] = nucleotide_based_ ? true : false;
+    std::string new_metadata_str = metadata.dump(4);
+
+    ret = tsk_table_collection_set_metadata(
+            p_tables, new_metadata_str.c_str(), (tsk_size_t)new_metadata_str.length());
+    if (ret != 0)
+        handle_error("tsk_table_collection_set_metadata", ret);
+
+    // As above, we maybe ought to edit the metadata schema adding our keys,
+    // but then comparing tables is a headache; see tskit#763
```

Comment:
It'd be nice if SLiM's top-level metadata could co-exist with other software's metadata. To do this, we can't just overwrite the metadata and metadata schema, we'd need to add to it.

Reply:
Ah, that.  Did my idea about allowing more than one schema for the top-level TS metadata get any traction?  That seems like the sensible thing to me.  You have, say, three schemas for the same metadata, each with their own required and optional keys, and the metadata should conform to all three.  Then we can just slap on a SLiM schema, rather than worrying about merging schemas together.  I forget when/where I proposed that; one of these multitudinous discussion threads somewhere in the ether.  :->

Reply:
Hm, interesting. This is equivalent to just adding the "SLiM" key to the list of properties... but, this is a discussion for later.